src/sirocco/parsing/yaml_data_models.py

```diff
@@ -102,30 +101,47 @@ def extract_merge_name(cls, data: Any) -> Any:
         return data
 
 
-class _WhenBaseModel(BaseModel):
-    """Base class for when specifications"""
-
-    before: datetime | None = None
-    after: datetime | None = None
-    at: datetime | None = None
-
-    @model_validator(mode="before")
-    @classmethod
-    def check_before_after_at_combination(cls, data: Any) -> Any:
-        if "at" in data and any(k in data for k in ("before", "after")):
-            msg = "'at' key is incompatible with 'before' and after'"
-            raise ValueError(msg)
-        if not any(k in data for k in ("at", "before", "after")):
-            msg = "use at least one of 'at', 'before' or 'after' keys"
+def select_when(spec: Any) -> When:
+    match spec:
+        case When():
+            return spec
+        case dict():
+            if not all(k in ("at", "before", "after") for k in spec):
+                msg = "when keys can only be 'at', 'before' or 'after'"
+                raise KeyError(msg)
+            if "at" in spec:
+                if any(k in spec for k in ("before", "after")):
+                    msg = "'at' key is incompatible with 'before' and after'"
+                    raise KeyError(msg)
+                return AtDate(**spec)
+            return BeforeAfterDate(**spec)
+        case _:
+            raise TypeError
+
+
+def select_target_cycle(spec: Any) -> TargetCycle:
+    match spec:
+        case TargetCycle():
+            return spec
+        case dict():
+            if tuple(spec.keys()) not in (("date",), ("lag",)):
```

Comment:
```suggestion
            if not (("date" in spec) ^ ("lag" in spec)):
```
Using XOR? I suppose there can be no other keys?

Reply:
It's actually making sure there are no other keys.